### ISSUE TITLE
Why are the integration tests written in golang?

### ISSUE THREAD

**Issue by davidnorthetal** [itchio/itch#2033]
Sorry to waste your time like this, but I'm wondering why using golang was preferred for the integration-tests as opposed to just JS/TS like https://electronjs.org/spectron ?

Thanks for any responses!
Dnorth

**Comment by fasterthanlime** [comment]
Hey David,

The short answer is reliability.

It might've been entirely my fault, but I just *could not* get spectron to run reliably. It invariably ran into a variant of:

  * chromedriver couldn't create a new session successfully
  * the tests just got stuck somewhere in the middle - maybe a response to a webdriver command was lost, I don't know exactly
  * chromedriver couldn't close the session cleanly (some electron processes were still running afterwards)

I tried for a really long time to get it working reliably, and finally just gave up.

Since electron is "just Chrome", and chrome "implements" WebDriver, and WebDriver is a "standard", and we use golang heavily at @itchio already (see https://github.com/itchio/butler and others) I forked [someone's golang webdriver client](https://github.com/bunsenapp/go-selenium) and added the parts we missed.

I found it much more reliable. Go needs to be compiled, but pure Go code (without cgo usage) is really fast to compile, it's negligible compared to the time it takes to bundle the application (typescript compiler + webpack - although that's gotten faster in the past few months, too).

All the retry logic (for session creation, waiting for an element to appear, clicking, etc.) is right there for me to inspect/tune as needed. Whenever a new webdriver command is needed, or a chromedriver workaround, it's a commit + `gvt update` away (gvt vendors go libraries, see https://github.com/FiloSottile/gvt) - whereas spectron depends on webdriverio, which in turns has its own set of dependencies, etc. - it would be much harder to ship just the fix we need.

Also, the thing I'm happiest about is that all the subprocesses (chromedriver, electron main/renderer/gpu etc.) are all in a single [Job Object](https://docs.microsoft.com/en-us/windows/desktop/procthread/job-objects) on Windows, with `JOB_OBJECT_LIMIT_KILL_ON_JOB_CLOSE` - which means they're always all closed, no matter how the job finishes (normally on test success, normally on test failure, abnormally on integration test timeout, abnormally on Gitlab CI timeout, etc.)

Not to mention that none of the spectron dependencies (& transitive dependencies) are pulled in the initial `npm install`. They're installed on-demand when `npm run integration tests` is run (it's only runnable by @itchio/core members, since it requires credentials for a test account to be set as environment variables). For infrequent OSS contributors, that's one less thing to worry about (since they're run in CI anyway).

Hopefully that answers you question!
